pkg/apis/flowcontrol/types.go

```diff
@@ -0,0 +1,278 @@
+	// +optional
+	metav1.ListMeta
+
+	// Items is a list of flow-schemas.
+	Items []FlowSchema
+}
+
+// FlowSchemaSpec describes how the flow-schema's specification looks like.
+type FlowSchemaSpec struct {
+	// PriorityLevelConfigurationRef can reference a PriorityLevelConfiguration in the cluster. If the reference cannot
+	// be resolved, the flow-schema will ignored and marked as invalid in its status.
+	PriorityLevelConfigurationRef PriorityLevelConfigurationRef
+	// MatchingPrecedence is used to choose among the FlowSchemas that match a given request.  The chosen
+	// FlowSchema is among those with the numerically lowest (which we take to be logically highest)
+	// MatchingPrecedence.  Each MatchingPrecedence value must be non-negative.
+	MatchingPrecedence int32
+	// DistinguisherMethod defines how to classify requests into certain flows. If not empty, the flow's identifier is computed
+	// according to specification of the distinguisher.
+	// +optional
+	DistinguisherMethod FlowDistinguisherMethod
+	// Rules holds all the RuleWithUser for this FlowSchema.
+	Rules []PolicyRuleWithSubject
+}
+
+// FlowDistinguisherMethod specifies the method of a flow distinguisher.
+type FlowDistinguisherMethod struct {
+	// Name is the name of flow distinguisher method
+	Name string
+}
+
+// PriorityLevelConfigurationRef contains information that points to the request-priority being used.
+type PriorityLevelConfigurationRef struct {
+	// Name is the name of resource being referenced
+	Name string
+}
+
+// PolicyRuleWithSubject prescribes a test that applies to a request to an apiserver. The test considers the subject
+// making the request, the verb being requested, and the resource to be acted upon.
+type PolicyRuleWithSubject struct {
+	// Subject is the user, group or serviceaccount that this rule cares about.
+	// +optional
+	Subject rbac.Subject
+	// Rule is the target resource or the subresource the rule cares about. APIGroups, Resources, etc.
+	// Required.
+	Rule rbac.PolicyRule
+}
+
+// FlowSchemaStatus represents the current state of a flow-schema.
+type FlowSchemaStatus struct {
+	// Current state of flow-schema.
+	Conditions []FlowSchemaCondition
+}
+
+// FlowSchemaCondition describes conditions for a flow-schema.
+type FlowSchemaCondition struct {
+	// Type is the type of the condition.
+	Type FlowSchemaConditionType
+	// Status is the status of the condition.
+	// Can be True, False, Unknown.
+	Status api.ConditionStatus
+	// Last time the condition transitioned from one status to another.
+	LastTransitionTime metav1.Time
+	// Unique, one-word, CamelCase reason for the condition's last transition.
+	Reason string
+	// Human-readable message indicating details about last transition.
+	Message string
+}
+
+// FlowSchemaConditionType is a valid value for FlowSchemaStatusCondition.Type
+type FlowSchemaConditionType string
+
+// FlowSchemaConditionStatus is a valid value for FlowSchemaStatusCondition.Status
+type FlowSchemaConditionStatus string
```

Comment:
We no longer need this type, nor the enumerated values of this type.

Reply:
nice catch